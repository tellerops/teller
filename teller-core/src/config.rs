use std::cmp::Ordering;
use std::path::PathBuf;
use std::{
    collections::{BTreeMap, HashMap},
    path::Path,
};

use fs_err as fs;
use serde_derive::{Deserialize, Serialize};
use teller_providers::config::{PathMap, ProviderCfg, KV};
use teller_providers::providers::ProviderKind;
use tera::{Context, Tera};

use crate::Result;

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct Config {
    pub providers: BTreeMap<String, ProviderCfg>,
}

#[derive(Serialize)]
pub struct RenderTemplate {
    pub providers: Vec<ProviderKind>,
}

fn apply_eqeq(config: &mut Config) {
    config.providers.iter_mut().for_each(|(_name, provider)| {
        provider.maps.iter_mut().for_each(|pm| {
            pm.keys.iter_mut().for_each(|(k, v)| {
                // THINK: replace with:
                // 1. templating: {{id}} (identity), {{snake_case}} (snake case it)
                // 2. other symbols: == id, ^^ capitalize, snake case __ lower snake case
                if v == "==" {
                    v.clone_from(k);
                }
            });
        });
    });
}

impl Config {
    /// Config from text
    ///
    /// # Errors
    ///
    /// This function will return an error if serialization fails
    pub fn with_vars(text: &str, vars: &HashMap<String, String>) -> Result<Self> {
        let rendered_text = Tera::one_off(text, &Context::from_serialize(vars)?, false)?;
        let mut config: Self = serde_yaml::from_str(&rendered_text)?;

        apply_eqeq(&mut config);

        Ok(config)
    }

    /// Config from text
    ///
    /// # Errors
    ///
    /// This function will return an error if serialization fails
    pub fn from_text(text: &str) -> Result<Self> {
        Self::with_vars(text, &HashMap::new())
    }

    /// Config from file
    ///
    /// # Errors
    ///
    /// This function will return an error if IO fails
    pub fn from_path(path: &Path) -> Result<Self> {
        Self::from_text(&fs::read_to_string(path)?)
    }

    /// Create configuration template file
    ///
    /// # Errors
    /// When could not convert config to string
    pub fn render_template(data: &RenderTemplate) -> Result<String> {
        let res: BTreeMap<String, ProviderCfg> = data
            .providers
            .iter()
            .map(|p| {
                (
                    format!("{p}_1"),
                    ProviderCfg {
                        kind: p.clone(),
                        maps: vec![PathMap::from_path("example/dev")],
                        ..ProviderCfg::default()
                    },
                )
            })
            .collect();

        let config = Self { providers: res };

        let a: String = serde_yaml::to_string(&config)?;
        Ok(a)
    }
}

#[derive(Debug, Clone, Serialize, Eq, PartialEq)]
pub struct Match {
    pub path: PathBuf,
    pub position: Option<(usize, usize)>,
    pub offset: usize,
    pub query: KV,
}

impl PartialOrd for Match {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Match {
    fn cmp(&self, other: &Self) -> Ordering {
        let query_cmp = self.query.cmp(&other.query);

        if query_cmp != Ordering::Equal {
            return query_cmp;
        }

        self.offset.cmp(&other.offset)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;
    #[test]
    fn load_config() {
        std::env::set_var("TEST_LOAD_1", "DEV");
        let config = Config::from_path(Path::new("fixtures/config.yml")).unwrap();
        assert_eq!(config.providers.len(), 2);
        assert_yaml_snapshot!(config);
    }

    #[test]
    fn can_render_template_config() {
        let data = RenderTemplate {
            providers: vec![ProviderKind::Inmem, ProviderKind::Dotenv],
        };

        let config = Config::render_template(&data).unwrap();
        assert_yaml_snapshot!(config);
    }
}
