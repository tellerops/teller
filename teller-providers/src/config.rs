use std::cmp::Ordering;
use std::collections::BTreeMap;

use serde_derive::{Deserialize, Serialize};

use crate::providers::ProviderKind;

fn is_default<T: Default + PartialEq>(t: &T) -> bool {
    t == &T::default()
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct ProviderCfg {
    #[serde(rename = "kind")]
    pub kind: ProviderKind,
    #[serde(rename = "options", skip_serializing_if = "Option::is_none")]
    pub options: Option<serde_json::Value>,
    #[serde(rename = "name", skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub maps: Vec<PathMap>,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub enum Sensitivity {
    #[default]
    None,
    Low,
    Medium,
    High,
    Critical,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct ProviderInfo {
    pub kind: ProviderKind,
    pub name: String,
}
#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct PathInfo {
    pub id: String,
    pub path: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct MetaInfo {
    pub sensitivity: Sensitivity,
    pub redact_with: Option<String>,
    pub source: Option<String>,
    pub sink: Option<String>,
}
#[derive(Serialize, Deserialize, Debug, Clone, Default, Eq, PartialEq)]
pub struct KV {
    pub value: String,
    pub key: String, // mapped-to key
    pub from_key: String,
    pub path: Option<PathInfo>, // always toplevel
    pub provider: Option<ProviderInfo>,
    pub meta: Option<MetaInfo>,
}

impl PartialOrd for KV {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for KV {
    fn cmp(&self, other: &Self) -> Ordering {
        if let (Some(provider), Some(other_provider)) = (&self.provider, &other.provider) {
            let provider_cmp = provider.kind.cmp(&other_provider.kind);
            if provider_cmp != Ordering::Equal {
                return provider_cmp;
            }
        }

        self.key.cmp(&other.key)
    }
}

impl KV {
    #[must_use]
    pub fn to_data(kvs: &[Self]) -> BTreeMap<String, String> {
        let mut data = BTreeMap::new();
        for kv in kvs {
            data.insert(kv.key.clone(), kv.value.clone());
        }
        data
    }

    #[must_use]
    pub fn from_data(
        data: &BTreeMap<String, String>,
        pm: &PathMap,
        provider: &ProviderInfo,
    ) -> Vec<Self> {
        // map all of the data found
        if pm.keys.is_empty() {
            data.iter()
                .map(|(k, v)| Self::from_value(v, k, k, pm, provider.clone()))
                .collect::<Vec<_>>()
        } else {
            // selectively map only keys from pathmap
            pm.keys
                .iter()
                .filter_map(|(from_key, to_key)| {
                    data.get(from_key).map(|found_val| {
                        Self::from_value(found_val, from_key, to_key, pm, provider.clone())
                    })
                })
                .collect::<Vec<_>>()
        }
    }
    #[must_use]
    pub fn from_value(
        found_val: &str,
        from_key: &str,
        to_key: &str,
        pm: &PathMap,
        provider: ProviderInfo,
    ) -> Self {
        Self {
            value: found_val.to_string(),
            key: to_key.to_string(),
            from_key: from_key.to_string(),
            path: Some(PathInfo {
                path: pm.path.clone(),
                id: pm.id.to_string(),
            }),
            provider: Some(provider),
            meta: Some(MetaInfo {
                sensitivity: pm.sensitivity.clone(),
                redact_with: pm.redact_with.clone(),
                source: pm.source.clone(),
                sink: pm.sink.clone(),
            }),
        }
    }
    #[must_use]
    pub fn from_literal(path: &str, key: &str, value: &str, provider: ProviderInfo) -> Self {
        Self {
            value: value.to_string(),
            key: key.to_string(),
            from_key: key.to_string(),
            path: Some(PathInfo {
                id: path.to_string(),
                path: path.to_string(),
            }),
            provider: Some(provider),
            ..Default::default()
        }
    }

    /// represents a KV without any source (e.g. created manually by a user, pending insert to
    /// one of the providers)
    #[must_use]
    pub fn from_kv(key: &str, value: &str) -> Self {
        Self {
            value: value.to_string(),
            key: key.to_string(),
            from_key: key.to_string(),
            ..Default::default()
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct PathMap {
    pub id: String,
    #[serde(rename = "protocol", skip_serializing_if = "Option::is_none")]
    pub protocol: Option<String>,
    #[serde(rename = "path")]
    pub path: String,
    #[serde(default, rename = "keys", skip_serializing_if = "is_default")]
    pub keys: BTreeMap<String, String>,
    #[serde(default, rename = "decrypt", skip_serializing_if = "is_default")]
    pub decrypt: bool,
    #[serde(default, rename = "sensitivity", skip_serializing_if = "is_default")]
    pub sensitivity: Sensitivity,
    #[serde(
        default,
        rename = "redact_with",
        skip_serializing_if = "Option::is_none"
    )]
    pub redact_with: Option<String>,
    #[serde(default, rename = "source", skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
    #[serde(default, rename = "sink", skip_serializing_if = "Option::is_none")]
    pub sink: Option<String>,
    // ignore population if optional + we got error
    #[serde(default, rename = "optional", skip_serializing_if = "is_default")]
    pub optional: bool,
}

impl PathMap {
    #[must_use]
    pub fn from_path(path: &str) -> Self {
        Self {
            path: path.to_string(),
            ..Default::default()
        }
    }
}
