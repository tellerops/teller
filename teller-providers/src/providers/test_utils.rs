use std::collections::{BTreeMap, HashMap};

use insta::{assert_debug_snapshot, with_settings};

use crate::config::{PathMap, KV};
use crate::{Error, Provider};

pub const ROOT_PATH_A: &str = "secret/development";
pub const ROOT_PATH_B: &str = "secret/multiple/app-1";
pub const ROOT_PATH_C: &str = "secret/multiple/app-2";
const PATH_A_KEY_1: &str = "db";
const PATH_A_KEY_2: &str = "log_level";
const PATH_A_KEY_3: &str = "app";
const PATH_A_VALUE_1: &str = "{\"DB_PASS\": \"1234\",\"DB_NAME\": \"FOO\"}";
const PATH_A_VALUE_2: &str = "DEBUG";
const PATH_A_VALUE_3: &str = "Teller";
const PATH_B_KEY_1: &str = "log_level";
const PATH_B_VALUE_1: &str = "DEBUG";
const PATH_C_KEY_1: &str = "foo";
const PATH_C_VALUE_1: &str = "bar";
const PATH_C_VALUE_1_UPDATE: &str = "baz";

pub struct ProviderTest {
    /// Adding the given prefix to all root path keys. you should use in case you want to change the root path key
    /// in case your provider is required a different path from the test case.
    /// In the all snapshots tests, the give value wan clean and you will not see it to aliment all the providers returns the sane response
    pub root_prefix: Option<String>,

    pub provider: Box<dyn Provider + Send + Sync>,
}

/// A struct representing a test suite for validating a Teller provider's functionality.
///
/// This struct encapsulates a set of test functions designed to validate the behavior of a Teller provider. It provides
/// methods for inserting, retrieving, updating, and deleting key-value pairs from the provider, allowing you to verify
/// the correctness of the provider's implementation.
///
/// # Example
/// ```
/// use tokio::test;
/// use crate::providers::test_utils;
/// use crate::Provider;
///
///
/// async fn sanity_test() {
///     let p = Box::new(super::Inmem::new(None).unwrap()) as Box<dyn Provider + Send + Sync>;
///
///     test_utils::ProviderTestBuilder::default().run(&p).await;
/// }
/// ```
impl ProviderTest {
    pub fn new(provider: Box<dyn Provider + Send + Sync>) -> Self {
        Self {
            root_prefix: None,
            provider,
        }
    }

    pub fn with_root_prefix(mut self, root_prefix: &str) -> Self {
        self.root_prefix = Some(root_prefix.to_string());
        self
    }

    pub async fn run(&self) {
        let path_tree = self.get_tree();

        self.validate_get_unexisting_key().await;
        self.validate_put(&path_tree).await;
        self.validate_get(&path_tree).await;
        self.validate_update().await;
        self.validate_delete().await;
        self.validate_delete_keys().await;
    }

    /// Returns a tree structure of test paths with associated key-value pairs.
    ///
    /// This function constructs a tree structure of test paths, where each path is associated with a vector of key-value pairs.
    /// The tree is represented as a `HashMap` where the keys are root paths, and the associated values are vectors of `KV` pairs.
    ///
    fn get_tree(&self) -> HashMap<&str, Vec<KV>> {
        HashMap::from([
            (
                ROOT_PATH_A,
                vec![
                    KV::from_literal(
                        "",
                        PATH_A_KEY_1,
                        PATH_A_VALUE_1,
                        self.provider.as_ref().kind(),
                    ),
                    KV::from_literal(
                        "",
                        PATH_A_KEY_2,
                        PATH_A_VALUE_2,
                        self.provider.as_ref().kind(),
                    ),
                    KV::from_literal(
                        "",
                        PATH_A_KEY_3,
                        PATH_A_VALUE_3,
                        self.provider.as_ref().kind(),
                    ),
                ],
            ),
            (
                ROOT_PATH_B,
                vec![KV::from_literal(
                    "",
                    PATH_B_KEY_1,
                    PATH_B_VALUE_1,
                    self.provider.as_ref().kind(),
                )],
            ),
            (
                ROOT_PATH_C,
                vec![KV::from_literal(
                    "",
                    PATH_C_KEY_1,
                    PATH_C_VALUE_1,
                    self.provider.as_ref().kind(),
                )],
            ),
        ])
    }

    fn get_key_path(&self, root_path: &str) -> String {
        self.root_prefix.as_ref().map_or_else(
            || root_path.to_string(),
            |prefix| format!("{prefix}{root_path}"),
        )
    }

    /// Validates that a Teller provider implementation returns an error when attempting to retrieve a non-existent key path.
    ///
    /// This function checks the behavior of the provided `Provider` implementation when trying to access an invalid key path
    /// by constructing a key path from a given root path and an invalid path segment. It then awaits the result and asserts
    /// that it is an error.
    async fn validate_get_unexisting_key(&self) {
        let res = self
            .provider
            .as_ref()
            .get(&PathMap::from_path(&format!("{ROOT_PATH_A}/invalid-path")))
            .await;
        assert!(res.is_err());
    }

    /// Validate that Teller provider successfully put all the tests tree map into the provider.
    async fn validate_put(&self, path_tree: &HashMap<&str, Vec<KV>>) {
        for (root_path, keys) in path_tree {
            let path_map = PathMap::from_path(&self.get_key_path(root_path));
            let res = self.provider.as_ref().put(&path_map, keys).await;
            assert!(res.is_ok());
            assert_debug_snapshot!(format!("[put-{}]", root_path.replace('/', "_"),), res);
        }
    }

    /// Validates that a Teller provider successfully stores a tree map of key-value pairs.
    ///
    /// This function tests the behavior of a Teller provider by attempting to put a collection of key-value
    /// into the provider and the function returns successfully response.
    /// Each put response function snapshot the function response struct
    async fn validate_get(&self, path_tree: &HashMap<&str, Vec<KV>>) {
        for root_path in path_tree.keys() {
            let res = self
                .provider
                .as_ref()
                .get(&PathMap::from_path(&self.get_key_path(root_path)))
                .await;
            assert!(res.is_ok());

            let mut res = res.unwrap();
            res.sort_by(|a: &KV, b| a.value.cmp(&b.value));

            with_settings!({filters => vec![
                (format!("{:?}", self.provider.as_ref().kind().kind).as_str(), "PROVIDER_KIND"),
                (format!("{:?}", self.provider.as_ref().kind().name).as_str(), "PROVIDER_NAME"),
                (format!("\".*{ROOT_PATH_A}").as_str(), format!("\"{ROOT_PATH_A}").as_str()),
                (format!("\".*{ROOT_PATH_B}").as_str(), format!("\"{ROOT_PATH_B}").as_str()),
                (format!("\".*{ROOT_PATH_C}").as_str(), format!("\"{ROOT_PATH_C}").as_str()),
            ]}, {
                assert_debug_snapshot!(
                format!("[get-{}]", root_path.replace('/', "_"),),
                res
            );
            });
        }
    }

    /// Validates the update operation on a Teller provider after inserting a tree structure.
    ///
    /// This function is intended to be executed after running the `validate_get` function to insert a tree structure into the
    /// Teller provider. It performs the following steps:
    ///
    /// 1. Updates a specific key-value pair with new values and expects a successful response from the provider.
    /// 2. Verifies that the updated value is reflected correctly in the provider.
    ///
    /// The function utilizes the provided `provider` to perform the update operation and snapshots the results for verification.
    async fn validate_update(&self) {
        // ** Update value and expected success response.
        let update_keys = vec![KV::from_literal(
            "",
            PATH_C_KEY_1,
            PATH_C_VALUE_1_UPDATE,
            self.provider.as_ref().kind(),
        )];
        let update_res = self
            .provider
            .as_ref()
            .put(
                &PathMap::from_path(&self.get_key_path(ROOT_PATH_C)),
                &update_keys,
            )
            .await;
        assert!(update_res.is_ok());
        assert_debug_snapshot!(
            format!("[put-update-{}]", ROOT_PATH_C.replace('/', "_"),),
            update_res
        );

        // ** Read the updated value and make sure that the value is updated.
        let read_after_update_res = self
            .provider
            .as_ref()
            .get(&PathMap::from_path(&self.get_key_path(ROOT_PATH_C)))
            .await;
        assert!(read_after_update_res.is_ok());

        with_settings!({filters => vec![
                (format!("{:?}", self.provider.as_ref().kind().kind).as_str(), "PROVIDER_KIND"),
                (format!("{:?}", self.provider.as_ref().kind().name).as_str(), "PROVIDER_NAME"),
                (format!("\".*{ROOT_PATH_A}").as_str(), format!("\"{ROOT_PATH_A}").as_str()),
                (format!("\".*{ROOT_PATH_B}").as_str(), format!("\"{ROOT_PATH_B}").as_str()),
                (format!("\".*{ROOT_PATH_C}").as_str(), format!("\"{ROOT_PATH_C}").as_str()),
            ]}, {
                assert_debug_snapshot!(
            format!("[get-after-update-{}]", ROOT_PATH_C.replace('/', "_"),),
            read_after_update_res
        );
            });
    }

    /// Validates the deletion of a key from a Teller provider.
    ///
    /// This function tests the behavior of a Teller provider by performing the following steps:
    ///
    /// 1. Deletes a specific key from the provider using the `del` method.
    /// 2. Verifies that the deletion operation is successful by checking the result for success.
    /// 3. Attempts to retrieve the deleted key and confirms that it returns an error, indicating that the key no longer exists.
    async fn validate_delete(&self) {
        let delete_res = self
            .provider
            .as_ref()
            .del(&PathMap::from_path(&self.get_key_path(ROOT_PATH_B)))
            .await;
        assert!(delete_res.is_ok());
        assert_debug_snapshot!(
            format!("[del-{}]", ROOT_PATH_B.replace('/', "_"),),
            delete_res
        );

        // ** Validate deletion key.

        let get_del_res = self
            .provider
            .as_ref()
            .get(&PathMap::from_path(&self.get_key_path(ROOT_PATH_B)))
            .await;

        assert!(matches!(
            get_del_res,
            Err(Error::NotFound { path: _, msg: _ })
        ));
        assert!(get_del_res.is_err());
    }

    /// Validates deletion specifics key from a Teller provider.
    ///
    /// This function tests the behavior of a Teller provider by performing the following steps:
    ///
    /// 1. Delete specifics keys from a path by using adding keys list.
    /// 2. Verifies that the deletion operation is successful by run `get` again on the same path and expecting that
    /// not see the deletion keys
    async fn validate_delete_keys(&self) {
        let mut path_path = PathMap::from_path(&self.get_key_path(ROOT_PATH_A));

        path_path.keys = BTreeMap::from([
            (PATH_A_KEY_2.to_string(), String::new()),
            (PATH_A_KEY_3.to_string(), String::new()),
        ]);

        let delete_keys_res = self.provider.as_ref().del(&path_path).await;
        println!("{delete_keys_res:#?}");
        assert!(delete_keys_res.is_ok());
        assert_debug_snapshot!(
            format!("[del-keys-{}]", ROOT_PATH_A.replace('/', "_")),
            delete_keys_res
        );

        let get_del_res = self
            .provider
            .as_ref()
            .get(&PathMap::from_path(&self.get_key_path(ROOT_PATH_A)))
            .await;

        with_settings!({filters => vec![
                (format!("{:?}", self.provider.as_ref().kind().kind).as_str(), "PROVIDER_KIND"),
                (format!("{:?}", self.provider.as_ref().kind().name).as_str(), "PROVIDER_NAME"),
                (format!("\".*{ROOT_PATH_A}").as_str(), format!("\"{ROOT_PATH_A}").as_str()),
            ]}, {
                assert_debug_snapshot!(
            format!("[get-del-keys-{}]", ROOT_PATH_A.replace('/', "_")),
            get_del_res
        );
            });
    }
}
