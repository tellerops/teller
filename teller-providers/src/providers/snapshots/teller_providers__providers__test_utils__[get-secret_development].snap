---
source: teller-providers/src/providers/test_utils.rs
expression: res
---
[
    KV {
        value: "Teller",
        key: "app",
        from_key: "app",
        path: Some(
            PathInfo {
                id: "",
                path: "secret/development",
            },
        ),
        provider: Some(
            ProviderInfo {
                kind: PROVIDER_KIND,
                name: PROVIDER_NAME,
            },
        ),
        meta: Some(
            MetaInfo {
                sensitivity: None,
                redact_with: None,
                source: None,
                sink: None,
            },
        ),
    },
    KV {
        value: "{\"DB_PASS\": \"1234\",\"DB_NAME\": \"FOO\"}",
        key: "db",
        from_key: "db",
        path: Some(
            PathInfo {
                id: "",
                path: "secret/development",
            },
        ),
        provider: Some(
            ProviderInfo {
                kind: PROVIDER_KIND,
                name: PROVIDER_NAME,
            },
        ),
        meta: Some(
            MetaInfo {
                sensitivity: None,
                redact_with: None,
                source: None,
                sink: None,
            },
        ),
    },
    KV {
        value: "DEBUG",
        key: "log_level",
        from_key: "log_level",
        path: Some(
            PathInfo {
                id: "",
                path: "secret/development",
            },
        ),
        provider: Some(
            ProviderInfo {
                kind: PROVIDER_KIND,
                name: PROVIDER_NAME,
            },
        ),
        meta: Some(
            MetaInfo {
                sensitivity: None,
                redact_with: None,
                source: None,
                sink: None,
            },
        ),
    },
]
